### ISSUE TITLE
`parJoin` hangs with ZIO

### ISSUE THREAD

**Issue by kaishh** [functional-streams-for-scala/fs2#1396]
Test case:
```scala
  def testCaseJoin[F[_]: Concurrent]: F[List[Int]] = {
    def one: F[Int]       = Sync[F].delay(1)
    val s: Stream[F, Int] = Stream.eval(one)
    val ss: Stream[F, Stream[F, Int]] = Stream.emits(List(s, s))
    ss.parJoin(2).compile.toList
  }

  testCaseJoin[scalaz.zio.interop.Task]
```

Details: on ZIO tracker - https://github.com/scalaz/scalaz-zio/pull/446#issuecomment-453678211

Gist: ZIO's `cancel` waits until the cancelled fiber actually dies unlike cats and monix's effect types, this creates an issue in `parJoin`/`interruptibleEval`:

```scala
  private[internal] def interruptibleEval[A](f: F[A]): F[Either[Either[Throwable, Token], A]] =
    interruptible match {
      case None => F.map(F.attempt(f)) { _.swap.map(Left(_)).swap }
      case Some(iCtx) =>
        F.map(
          iCtx.concurrent
            .race(iCtx.deferred.get, F.attempt(iCtx.concurrent.uncancelable(f)))) {
          case Right(result) => result.leftMap(Left(_))
          case Left(other)   => Left(other)
        }
}
```

If `iCtx.deferred.get` wins the race, the right task fiber will be cancelled - but it's uncancelable, and apparently doesn't finish, so ZIO locks. I don not know why the `f` might hang forever in `parJoin` - it's possible that ZIO's cats-effect implementation is broken in some way and causes this. There is also a possiblity `parJoin` has always silently leaked fibers, but went unnoticed until now.

@SystemFw @mpilquist Thoughts?

**Comment by SystemFw** [comment]
> ZIO's cancel waits until the cancelled fiber actually dies unlike cats and monix's effect types

Actually both `IO` and `Monix.Task` also wait for the fiber to die, so I would also look into

>  it's possible that ZIO's cats-effect implementation is broken in some way and causes this. 

That being said, the `interruptibleEval` is a known issue, see https://github.com/functional-streams-for-scala/fs2/issues/1140#issuecomment-437934733

I have every intention to fix that ofc, it's just a matter of free time/complexity

**Comment by mpilquist** [comment]
Fixed by #1419
